### PR TITLE
Data Mapper #34

### PR DESCRIPTION
**Pull request by inbravo**
A basic data-mapper implementation based on discussion on #34

### CODE REVIEW

**Comment by iluwatar** [comment]
Review comments:
- `StudentFirstMapper` and `StudentSecondMapper` seem very similar. Can we drop one of them to simplify the example? You wouldn't need to process command line parameters in `App` after the change.
- In `App` the conditional use of `Logger` is confusing. Could we just output the debug log without testing any conditions?
- `DataMapperException` seems to forward the call to `super` in most of its methods. In this case you can leave out the implementation because it's the default behavior.

File: data-mapper/src/main/java/com/iluwatar/datamapper/StudentDataMapper.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+ * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.iluwatar.datamapper;
+
+import java.util.Optional;
+
```
Comment (iluwatar) [inline data-mapper/src/main/java/com/iluwatar/datamapper/StudentDataMapper.java]: Add JavaDoc

File: data-mapper/src/main/java/com/iluwatar/datamapper/DataMapperException.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
+/**
+ * The MIT License Copyright (c) 2016 Amit Dixit
```
Comment (iluwatar) [inline data-mapper/src/main/java/com/iluwatar/datamapper/DataMapperException.java]: Please keep the license header consistent

**Comment by iluwatar** [comment]
@inbravo you have my review comments. Please notify me once you've addressed them.

**Comment by inbravo** [comment]
- `StudentSecondDataMapper ` is dropped and `StudentFirsDataMapper ` is renamed to `StudentDataMapperImpl` for better readability.
-  Conditional usage of `logger` in App class is discarded.
-  `DataMapperException` is forwarding calls to `RuntimeException` which is a core Exception class but not any reference from implementation. This is helpful in detailing users about removing all data-mapper exception dependency from code. Please review again.
- Student equality test is added at `StudentTest`
- Your last point is not clear to me; please share more information on it.

I hope just a final review is required.

File: data-mapper/src/main/java/com/iluwatar/datamapper/DataMapperException.java (hunk 2 above)
Comment (inbravo) [inline data-mapper/src/main/java/com/iluwatar/datamapper/DataMapperException.java]: Done!

**Comment by iluwatar** [comment]
@inbravo took another look. Here are my comments:
- `DataMapperException` defines additional constructors that are never used. These should be removed.
- You should add `data-mapper` as a module to the main [pom.xml](https://github.com/iluwatar/java-design-patterns/blob/master/pom.xml) Look for `<modules>` declaration and see how the other modules have been added.
- (optional improvement) Since this is a database related pattern it would be great to see actual real database underneath.

Please comment when you're ready for another look.

**Comment by inbravo** [comment]
Parent info is required, i am giving this, please review, build is failed

	<parent>
		<groupId>com.iluwatar</groupId>
		<artifactId>java-design-patterns</artifactId>
		**<version>1.7.0</version>**
	</parent>

**Comment by iluwatar** [comment]
Looks like you have the wrong version number. Try this:
```
<parent>
    <groupId>com.iluwatar</groupId>
    <artifactId>java-design-patterns</artifactId>
    <version>1.12.0-SNAPSHOT</version>
</parent>
```

**Comment by inbravo** [comment]
- Unused constructors from `DataMapperException` discarded
- Main build file updated for addition of new module
- My initial cut was based on DB based assumption; you are absolutely correct it should be DB based, let me see if i can do this ...

**Comment by inbravo** [comment]
This pattern is ready for release; you can plan out ....

**Comment by iluwatar** [comment]
@inbravo I've merged the initial implementation. Looking forward to the real database implementation though.

**Pull request merged**